website/server/libs/groupTasks.js

```diff
@@ -1,24 +1,73 @@
 import * as Tasks from '../models/task';
+import {model as Groups} from '../models/group';
+import {model as Users} from '../models/user/index';
 
 const SHARED_COMPLETION = {
   default: 'recurringCompletion',
   single: 'singleCompletion',
   every: 'allAssignedCompletion',
 };
 
-async function _completeMasterTask (masterTask) {
-  masterTask.completed = true;
+async function _completeOrUncompleteMasterTask (masterTask, completed) {
+  masterTask.completed = completed;
   await masterTask.save();
 }
 
-async function _deleteUnfinishedTasks (groupMemberTask) {
-  await Tasks.Task.deleteMany({
-    'group.taskId': groupMemberTask.group.taskId,
-    $and: [
-      {userId: {$exists: true}},
-      {userId: {$ne: groupMemberTask.userId}},
-    ],
-  }).exec();
+async function _updateAssignedUsersTasks (masterTask, groupMemberTask) {
+  if (groupMemberTask.type == 'todo') {
+    if (groupMemberTask.completed) {
+      // The task was done by one person and is removed from others' lists
+      await Tasks.Task.deleteMany({
+        'group.taskId': groupMemberTask.group.taskId,
+        $and: [
+          {userId: {$exists: true}},
+          {userId: {$ne: groupMemberTask.userId}},
+        ],
+      }).exec();
+    } else {
+      // The task was uncompleted by the group member and should be recreated for assignedUsers
+      let group = await Groups.findById(masterTask.group.id);
+      let userList = [];
+      masterTask.group.assignedUsers.forEach(userId => {
+        let query = {_id: userId};
+        userList.push(query);
+      });
+      // @REVIEW There has to be a better way to do this
+      // async arrow function callback to the find and/or a forEach resulted in Mongoose ParallelSaveErrors
+      let assignedUsers = await Users.find({
+          $or: userList
+        }).exec();
+      for (let i=0; i < assignedUsers.length; i++) {
+        let promises = [];
+        promises.push(group.syncTask(masterTask, assignedUsers[i]));
+        promises.push(group.save());
+        await Promise.all(promises);
+      }
+    }
+  } else {
+    // Complete or uncomplete the task on other users' lists
+    await Tasks.Task.find({
+      'group.taskId': groupMemberTask.group.taskId,
+      $and: [
+        {userId: {$exists: true}},
+        {userId: {$ne: groupMemberTask.userId}}
+      ]},
+      function (err, tasks) {
+        // @REVIEW How does Habitica handle errors?
+        if (err) return;
```

Comment:
WIP. How does Habitica handle errors?

Reply:
This should be converted to a promise (see the comment above about using `exec()`) and in that case errors will be handled automatically (unless they need some special handling but it doesn't seem the case)

Reply:
Makes sense.

Reply:
Pushed rework.